### ISSUE TITLE
CUDA 10.2 faster than 11 on older hardware

### ISSUE THREAD

**Issue by hermancollin** [ivadomed/ivadomed#1141]
While testing #1129 on the ADS side on `bireli`, we found this weird behavior. Downgrading CUDA from 11.1 to 10.2 speeds up inference (almost twice as fast).

`bireli` has a [GeForce GTX TITAN X](https://www.nvidia.com/en-us/geforce/graphics-cards/geforce-gtx-titan-x/specifications/) (2014). This is the output of `time` on our quick integrity test on bireli:

>CUDA 11.1:
>```
>real    0m16.210s
>user    0m52.484s
>sys     0m3.251s
>```
>CUDA 10.2:
>```
>real    0m9.065s
>user    0m6.337s
>sys     0m1.478s
>```

The performance is different with newer hardware, e.g. here is on `romane`, which has [RTX A6000](https://www.nvidia.com/en-us/design-visualization/rtx-a6000/) (2020):
>CUDA 11.1:
>```
>real    0m8.021s
>user    0m12.308s
>sys     0m2.939s
>```

My guess:
> so, no it's not as slow. I'm guessing this is because `bireli`'s gpus are older. It seems a lot of people reported slower inference time with CUDA 11 in comparison with 10.2: https://github.com/pytorch/pytorch/issues/47908
> CUDA 11.1 is shipped with cudnn 8 whereas CUDA 10.2 is used from an older cudnn 7. It could be related to the cudnn version and if `torch.backends.cudnn.benchmark == True`, which seems to be the case for ivadomed, e.g. here:
https://github.com/ivadomed/ivadomed/blob/7b76bf81a025cde3096fd1d686d6f3c0b8ce8f02/ivadomed/main.py#L28 
according to [this issue](https://github.com/pytorch/pytorch/issues/51044)

However, setting `cudnn.benchmark = True` to False in main.py and testing.py, I wasn't able to observe a meaningful time difference but maybe I messed something up.

With this being said, this is low-priority stuff IMO because it shouldn't have any impact on newer hardware. Just wanted to let everyone know about this.

**Comment by uzaymacar** [comment]
Thanks for letting us know! Isn't it also problematic that `romane` on CUDA 11.2 with RTX A6000 is NOT faster than `bireli` with CUDA 10.2? I believe there is a considerable slow down in `romane` as well (also tagging @naga-karthik).

**Comment by hermancollin** [comment]
> Thanks for letting us know! Isn't it also problematic that `romane` on CUDA 11.2 with RTX A6000 is NOT faster than `bireli` with CUDA 10.2? I believe there is a considerable slow down in `romane` as well (also tagging @naga-karthik).

You are absolutely right. To be fair, this benchmark wasn't extremely rigorous so further testing would be required. 

The performance I timed with cuda 10.2 was very surprising to me because we are actually not used to processing data as fast (we were using cuda 11 on bireli). I was not able to test this cuda version on romane because I'm not even sure it's compatible with romane's newer hardware.

**Comment by hermancollin** [comment]
Actually, using `python -m cProfile` and something like [flameprof](https://github.com/baverman/flameprof), we could generate nice flamegraphs to see if something stands out, comparing the profiles obtained on both clusters.

Another more targetted approach would be to use [NVIDIA's Deep Learning profiler](https://docs.nvidia.com/deeplearning/frameworks/dlprof-user-guide/) (which I never used but it looks pretty neat)

**Comment by hermancollin** [comment]
> This is an upstream bug to be resolved in PyTorch 1.11.0. Look at [these](https://github.com/pytorch/pytorch/issues/47908#issuecomment-1146884119) updated comments upstream. We are yet to update ivadomed to support PyTorch 1.11.0 but feel free to use `torch~=1.11.0` instead of `torch~=1.8.1` by updating it in `requirements.txt` and report any bugs.

I believe the issue here is that even tho I can downgrade cuda on bireli, romane should still be way faster regardless of the cuda version used because its gpus are much newer

**Comment by kanishk16** [comment]
> I believe the issue here is that even tho I can downgrade cuda on bireli, romane should still be way faster regardless of the cuda version used because its gpus are much newer

I guess there are a lot of factors associated with it, even batch_size while inference makes a diff ([comment upstream](https://github.com/pytorch/pytorch/issues/47908#issuecomment-879108708))
